yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,785 @@
+      return { preimage: lowPublicDataPreimage, index: index, update: alreadyPresent };
+    }
+
+    // We look for the low element by bouncing between our local indexedUpdates map or the external DB
+    // The conditions we are looking for are:
+    // (1) Exact Match: curr.nextKey == key (this is only valid for public data tree)
+    // (2) Sandwich Match: curr.nextKey > key and curr.key < key
+    // (3) Max Condition: curr.next_index == 0 and curr.key < key
+    // Note the min condition does not need to be handled since indexed trees are prefilled with at least the 0 element
+    let found = false;
+    let curr = minPreimage.preimage as T;
+    let result: PreimageWitness<T> | undefined = undefined;
+    // Temp to avoid infinite loops - the limit is the number of leaves we may have to read
+    const LIMIT = 2n ** BigInt(getTreeHeight(treeId)) - 1n;
+    let counter = 0n;
+    let lowPublicDataIndex = minPreimage.index;
+    while (!found && counter < LIMIT) {
+      if (curr.getKey() === bigIntKey) {
+        // We found an exact match - therefore this is an update
+        found = true;
+        result = { preimage: curr, index: lowPublicDataIndex, update: true };
+      } else if (curr.getKey() < bigIntKey && (curr.getNextKey() === 0n || curr.getNextKey() > bigIntKey)) {
+        // We found it via sandwich or max condition, this is a low nullifier
+        found = true;
+        result = { preimage: curr, index: lowPublicDataIndex, update: false };
+      }
+      // Update the the values for the next iteration
+      else {
+        lowPublicDataIndex = curr.getNextIndex();
+        if (this.hasLocalUpdates(treeId, lowPublicDataIndex)) {
+          curr = this.getIndexedUpdates(treeId, lowPublicDataIndex)!;
+        } else {
+          const preimage: IndexedTreeLeafPreimage = (await this.treeDb.getLeafPreimage(treeId, lowPublicDataIndex))!;
+          curr = preimage as T;
+        }
+      }
+      counter++;
+    }
+    // We did not find it - this is unexpected
+    if (result === undefined) {
+      throw new Error('No previous value found or ran out of iterations');
+    }
+    return result;
+  }
+
+  /**
+   * This hashes the preimage to a field element
+   */
+  hashPreimage<T extends TreeLeafPreimage>(preimage: T): Fr {
+    // Watch for this edge-case, we are hashing the key=0 leaf to 0.
+    // This is for backward compatibility with the world state implementation
+    if (preimage.getKey() === 0n) {
+      return Fr.zero();
+    }
+    const input = preimage.toHashInputs().map(x => Fr.fromBuffer(x));
+    return poseidon2Hash(input);
+  }
+}
+
+/****************************************************/
+/****** Some useful Structs and Enums **************/
+/****************************************************/
+enum TreeType {
+  LEAF,
+  BRANCH,
+}
+
+type Leaf = {
+  tag: TreeType.LEAF;
+  value: Fr;
+};
+type Branch = {
+  tag: TreeType.BRANCH;
+  leftTree: Tree;
+  rightTree: Tree;
+};
```

Comment:
I personally would still rename this "Node" as "branch" to me would be either the left path or the right path, not a node containing both